Add a feature flag to the raw CsgDifferenceOp to immediately prune degenerate tiles.

jmlait opened the pull request:
When the same tile subtracted from itself a background tile is created.
greatly reduces the generated ghost borders when similar objects are
differenced.

Provided for discussion.   Is a template parameter correct way for this?  I want this in VDB Combine as a toggle parameter, so do we also add parameters to higher level functions?  This also can extend to Union/Intersect.

And is this even a good thing?

danrbailey commented on openvdb/openvdb/tools/Merge.h:
```diff
@@ -1158,30 +1161,40 @@ bool CsgDifferenceOp<TreeT>::operator()(LeafT& leaf, size_t) const
         const ValueT& aValue = leaf.getValue(i);
-        ValueT bValue = math::negative(mergeLeaf->getValue(i));
+        ValueT bValue = mergeLeaf->getValue(i);
+        if (PruneCancelledTiles)
```
Probably worth doing `PruneCancelledTiles && allequal` here so the first non-match disables subsequent value checks.

jmlait replied:
I'd sort of hope the compiler would deal with
allequal &= foo;
implying
if (allequal) allequal = foo;
for boolean foo/allequal and no side effects on foo...  We also don't get to avoid value checks because we have to load aValue and bValue in any case, so it is mostly if the cost of the branch is worth eliding a comparison and logic operation...   My random estimate is branches are more expensive, so the branch free seems best to me, but with the potential the compiler will ignore this at its discretion.

danrbailey replied:
You're probably right about that.

danrbailey reviewed (commented):
> When the same tile subtracted from itself a background tile is created. greatly reduces the generated ghost borders when similar objects are differenced.
> 
> Provided for discussion. Is a template parameter correct way for this? I want this in VDB Combine as a toggle parameter, so do we also add parameters to higher level functions? This also can extend to Union/Intersect.
> 
> And is this even a good thing?

I don't see an ideal solution here as we can't easily modify the constructors due to the tag dispatch class. I think the simplest approach is to just add a runtime modifier such as:

`CsgDifferenceOp::setPruneCancelledTiles(true);`

then if we want to maintain the performance of the non-pruned leaf method (which we most likely do), then we either add a private member function that is templated to optimize out the value equality checking or for readability what I'd probably do in this case is to add a lambda for the `aValue < bValue` clause then live with a little code duplication for the rest of the loop. That at least avoids performing redundant template instantiation for the other methods that don't use this parameter.

jmlait commented:
> I don't see an ideal solution here as we can't easily modify the constructors due to the tag dispatch class. I think the simplest approach is to just add a runtime modifier such as:
> 
> `CsgDifferenceOp::setPruneCancelledTiles(true);`
> 
> then if we want to maintain the performance of the non-pruned leaf method (which we most likely do), then we either add a private member function that is templated to optimize out the value equality checking or for readability what I'd probably do in this case is to add a lambda for the `aValue < bValue` clause then live with a little code duplication for the rest of the loop. That at least avoids performing redundant template instantiation for the other methods that don't use this parameter.

This is a good idea....  I think, however, the correct solution is just to duplicate the loop.  We have a 512:1 amortization rate with the leaf iteration, and it is just a few lines duplicated.  I think setting up the template or other approach would generate more lines of code than we end up actually repeating here.  I also like that the allequal stack variable can be skipped in the other loop making things clearer.

jmlait commented:
I've added support for union/intersection, though it is somewhat difficult to setup these cases in practice.  Two touching cubes tend to convert slightly differently so don't trigger it unless I round off distance values artificially.

Also switched to the member variable to avoid template explosion.

jmlait commented on openvdb/openvdb/tools/Merge.h:
```diff
@@ -966,12 +977,35 @@ bool CsgUnionOrIntersectionOp<TreeT, Union>::operator()(LeafT& leaf, size_t) con
+                // If two diffed tiles have the same values of opposite signs,
+                // we know they have both the same distances and gradients.
+                // Thus they will cancel out.
+                if (Union) { leaf.fill(math::negative(this->background()), false); }
+                else { leaf.fill(this->background(), false); }
```
I don't like this->background(), but this was done at the start of this method!
const ValueT background = Union ? this->background() : -this->background();

danrbailey replied:
I don't understand, why can't you just do:
```
leaf.fill(background, false);
```
?

jmlait replied:
That background is opposite sign to what we want.   So it could be
leaf.fill(math::negative(background), false);
but that would be rather confusing... I like the background always meaning positive background.

danrbailey replied:
Ah yes, makes sense.

danrbailey commented:
This looks fine to me, though there is perhaps a little ambiguity on what whether `prune=false` will override `prunecancelledtiles=true`. There's a missing explicit template instantiation parameter that is causing the CI to fail. Also, minor quibble, I think we tend to prefer lower camel case for parameters (ie `pruneCancelledTiles` instead of `prunecancelledtiles`).

danrbailey reviewed (approved):
Looks good to me! It appears that the CI is now failing for unrelated Windows reasons.
